### DISCUSSION TITLE
Configuration options per area (SPI)

### DISCUSSION THREAD

**Discussion by martin-kanis** [keycloak/keycloak#13623] [Ideas]
Currently, when we want to configure a factory property, we specify it in `--<spi>-<provider>-<property>=value` format i.e `--spi-authentication-sessions-infinispan-auth-sessions-limit=100`. For the new storage this will become `--spi-authentication-sessions-map-auth-sessions-limit=100`
Would it make sense to introduce a configuration option per area (SPI) that will apply to whatever implementation the given area is implemented by? This would simplify configuration because users won't have to think about what implementation is used for the given area. They will focus just on the property itself. Also this would simplify documentation where we can provide just generic way how to configure a property independent of used provider.

**Comment by sguilhen** [comment]
So, for some SPIs we have configuration options that are not provider-specific, and instead of having a property variant for every provider we would have a single property at the SPI level, right? Using your example, we would have something like `--spi-authentication-sessions-auth-sessions-limit=100` instead of one for `infinispan` and another one for `map` that essentially configure the same aspect.

As these would be considered SPI-level properties, each provider implementation basically re-uses the same property when it's defined instead of having its own version, which would also eliminate the need to translate the generic property into provider-specific ones.

**Reply by mhajas** [reply]
This sounds like a good idea. It seems like it could be easy to implement. Here:
https://github.com/keycloak/keycloak/blob/e44cea587fe97d73b83de2ea8e943cf079ce95ee/services/src/main/java/org/keycloak/services/DefaultKeycloakSessionFactory.java#L280-L291

we can merge spi-level scope with provider-specific scope (with higher priority on provider-specific values).

However, I am not sure how this would work for Amphibian factories.

**Comment by pedroigor** [comment]
I would say that the situation is temporary and should not be a problem once we drop the legacy store.

For options that we are sure we want to expose to users, we should have them mapped accordingly like `storage-auth-sessions-limit` (the name is just an example).

From a wider perspective, provider options (`spi-*`) are provider-specific and we can't assume an option is available between multiple implementations of a single SPI. The overall idea here is to:

* Make sure single-implementation SPIs have their options exposed and mapped accordingly and if needed/wanted
* Allow exposing SPI options from providers. Like we do here for some SPIs https://www.keycloak.org/server/all-provider-config.
